Finalization of Events implementation

vidartf opened the pull request:
Implements Events across the existing code-base, to replace existing varying callback functionality. This also adds an arg to `Model.multifit` to prevent plotting during fitting.

vidartf committed:
Separate out Event nargs resolution

Make it a separate function to ease overloading.

vidartf committed:
Improved markers' dynamics

vidartf committed:
Create named container in Events

Makes Event book-keeping cleaner (don't rely on `__dict__`!), but still
makes it clear that the events should be accessed through methods on
`Events` by making the container protected.

vidartf committed:
Add missing `Signal.update_plot`

Dependency on `Signal.update_plot` was merged earlier.

vidartf committed:
Code cleanup

PEP8++

vidartf committed:
Replace Parameters connections with Events

vidartf committed:
Create events for DataAxis/AxesManager

Create events for all triggers in DataAxis/AxesManager that are used
outside. This keeps the internal traits implementation, but isolates it
to these two classes, making future changes a little bit easier. The
change also improves even timings, as now all external processing
happens after the internal processing, which is really important for
e.g. having AxesManager finish `_update_attributes()` before updating
plots etc.

vidartf committed:
Improve general code by AxesManager events

Removed dummy-function in plotting code, and added `interactive_plot`
option to `Model.multifit`.

vidartf commented:
You'll have to merge master/0.8.x to get a neat diff. New code is 8 last commits.

vidartf commented:
Let me know if there are any more existing callback structures that needs to be replaced as well.

vidartf committed:
Fix introduced errors

vidartf committed:
DOC: Events docstrings

Added and improved events' docstrings.

francisco-dlp commented:
Same problem here, diff too long.

vidartf commented:
I'll rebase here as well. I think it might be because the internal branch hell I've been having here, moving stuff back and forth.

vidartf commented:
Pushing clean rebase now, but I have some changes I want to do to the proposed changes, so no need to review yet.

francisco-dlp commented:
ok, thanks.

vidartf committed:
Created `AxesManager.navigated` event

This event more correctly reflect what should be notified to users of
`connect`.

Also updated `AxesManager.axes_changed` such that it triggers for all
relevant events except navigation.

vidartf committed:
Simpler, better dummy CM

vidartf committed:
Added `__repr__` to `Event`/`Events`

vidartf committed:
Added `Events.update()`

Useful for when several `Events` structures need to be combined. Prefix
as an optional argument allows to avoid name collisions that would
overwrite unintentionally

vidartf committed:
Cleanup

vidartf committed:
Make `Model.suspend_update()` CM, remove resume

vidartf committed:
Fixed overly trigger-happy event

vidartf commented:
OK, this should be good for review now. I've purposefully been aggressive in making api changes, so please let me know what you think of it. Some of these changes could also be extracted to separate PRs to hyperspy master for review once hyperspy/hyperspy#423 is merged.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -82,6 +83,10 @@ def __init__(self,
         super(DataAxis, self).__init__()
+        self.events = Events()
+        self.events.axis_changed = Event()
+        self.events.navigated = Event()
```
Why not `index_changed` instead?

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -469,6 +475,9 @@ class AxesManager(t.HasTraits):
         super(AxesManager, self).__init__()
+        self._events = Events()
+        self._events.axes_changed = Event()
+        self._events.navigated = Event()
```
`axes_index_changed`?

vidartf replied:
`indices_changed` in latest.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -478,11 +487,22 @@ def __init__(self, axes_list):
-        self.on_trait_change(self._update_attributes, '_axes.index')
-        self.on_trait_change(self._update_attributes, '_axes.size')
+        self.on_trait_change(self._on_navigate, '_axes.index')
+        self.on_trait_change(self._on_nav_axes_changed, '_axes.slice')
```
`view_changed`?

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -478,11 +487,22 @@ def __init__(self, axes_list):
+        self.on_trait_change(self._on_navigate, '_axes.index')
+        self.on_trait_change(self._on_nav_axes_changed, '_axes.slice')
+        self.on_trait_change(self._on_axes_size_changed, '_axes.size')
+        self.on_trait_change(self._events.axes_changed.trigger, '_axes.scale')
```
`axes_calibration_changed`?

vidartf replied:
See Event docstring for reasoning.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -478,11 +487,22 @@ def __init__(self, axes_list):
+        self.on_trait_change(self._on_nav_axes_changed, '_axes.slice')
+        self.on_trait_change(self._on_axes_size_changed, '_axes.size')
+        self.on_trait_change(self._events.axes_changed.trigger, '_axes.scale')
+        self.on_trait_change(self._events.axes_changed.trigger, '_axes.offset')
```
`axes_calibration_changed`?

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -655,6 +675,18 @@ def _append_axis(self, *args, **kwargs):
+
+    def _on_nav_axes_changed(self):
+        self._update_attributes()
+        self._events.axes_changed.trigger()
```
`view_changed`?

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -655,6 +675,18 @@ def _append_axis(self, *args, **kwargs):
+
+    def _on_axes_size_changed(self):
+        self._update_attributes()
+        self._events.axes_changed.trigger()
```
`axes_size_changed`? Otherwise, if you don't want to be that specific, just connecting to the `axis_changed` events should do.

vidartf replied:
Connecting to all the individual `DataAxis` events will potentially trigger multiple time per change, while the single `AxesManager` event will hopefully only trigger once by intelligent use of `suppress`.

Rule-of-thumb: If you are interested in the event on one specific axis, connect to that speciic axes. If you are interested in an event on all axes, use the ones on the axes manager.

francisco-dlp commented on hyperspy/component.py:
```diff
@@ -113,7 +115,8 @@ class Parameter(t.HasTraits):
         self._twins = set()
-        self.connected_functions = list()
+        self.events = Events()
+        self.events.parameter_changed = Event()
```
`parameter_value_changed`? or `value_changed`?

francisco-dlp commented on hyperspy/component.py:
```diff
@@ -180,16 +183,10 @@ def __len__(self):
         return self._number_of_elements
 
     def connect(self, f):
-        if f not in self.connected_functions:
```
Probably we should deprecate this method for removal in 0.10 and avoid using it internally.

francisco-dlp commented on hyperspy/component.py:
```diff
@@ -180,16 +183,10 @@ def __len__(self):
-                self.twin.connect(f)
+        self.events.parameter_changed.connect(f)
 
     def disconnect(self, f):
```
Same as above.

francisco-dlp commented on hyperspy/events.py:
```diff
@@ -97,14 +111,46 @@ def suppress(self):
+            object.__delattr__(name)
+
+    def __repr__(self):
+        return "hyperspy.events.Events: " + str(self._events)
```
the string should be sourrounded by < >

francisco-dlp commented on hyperspy/misc/utils.py:
```diff
@@ -111,6 +112,11 @@ def unfold_if_multidim(signal):
 
 
+@contextmanager
+def dummy_context_mgr(*args, **kwargs):
```
`dummy_context_manager`? (More readable, not much longer.)

francisco-dlp commented on hyperspy/models/model1D.py:
```diff
@@ -289,49 +289,30 @@ def update_plot(self, *args, **kwargs):
+        for component in self:
+            evs.update(component.events, prefix=component.name)
+            for parameter in component.parameters:
+                evs.update(parameter.events, prefix=component.name + '.' +
```
The suppress all parameter and component events. We should be more specific.

francisco-dlp commented:
The new `Events` feature have a number of drawbacks:
- IPython tab-completion does not work with events.
- Overloading Python methods makes it hard to the user to understand what's going on. I would say that we should only do this when there are important benefits to be gained, but I'm not convinced that this is the case here.
- The `update` method is designed to add events from other `Event` instances for a given `Event` instance. You use it to ease the suppression of events from different containers. I think that adding all axes events to the `AxesManager` unnecessarily clutter the `AxesManager` events. I see the point of easily suppressing events in different containers but, why not solving this by writing a function that, given a list of `Events` instances and/or `Event` instances, returns a context manager to suppress them all?

Finally, I think that it is important to document what each event does, including the keyword arguments that will be sent when triggered. What about doing this in each `Event` instance docstring?

vidartf commented:
To start at the end: I agree with giving each Event a docstring. I think the easiest way to do this is to add an optional argument in the Event constructor, e.g.

``` python
signal.events.data_changed = Event("""
    Event that triggers when the data has changed and is ready for consumption by any
    process that depend on it as input. Plotted signals automatically connect this Event 
    to its `Signal.plot()`.

    Event arguments:
        signal: The signal that owns the data.
    """)
```

I'll add suggestions for docstrings for the events that I've implemented so far, and then we can use those as a basis for further discussion about behavior.

vidartf commented:
Adding IPython tab-completion is just a matter of implementing `Events.__dir__()` if we want to keep the current structure. If we're afraid that too much black magic will scare away other contributors, then I can try to comment the relevant bits (docstrings?).

`Events.update` is a general-purpose method for manipulating `Event` containers. Currently the only use is as you say. The reason I implemented it is because it will need to be a part of any solution that wants to suppress several `Events` at once, at least until we no longer support python less than 3.3 (see http://stackoverflow.com/questions/3024925/python-create-a-with-block-on-several-context-managers). Whether the solution in `AxesManager` is the best, I'm not sure. I see your concerns, and tried to come up with a solution myself:
- We can have a function `AxesManager.suppress_all_events()` (other name suggestions are welcome), but that is not in the intuitive place (`AxesManager.events.<something>`).
- We can monkey-patch a function on to `AxesManager.events` (by setattr/getattr changes, this is possible), e.g. `AxesManager.events.suppress_all_events()`. However, that is maybe not so ideal as the user might expect this to be available for all `Events`, and it makes the code a little harder to navigate.
- We can add some changes to `Events`. E.g. `Events.children = []` + `Events.suppress_hierarchy()` that suppresses itself + calls `suppress_hierarchy` on any `Events` in children (this can then propagate further).

In the end, I ended up trying the simplest first to see what thoughts you had :) Among the three suggestions listed above, I think I would prefer the `suppress_hierarchy` solution.

vidartf commented:
PS: The `suppress_hierarchy` solution above, or any other similar future changes to `Events`, require some black magic to sort out which attributes are events.

An alternative is of course to extract all the black magic to a separate utility class stored somewhere else, and then inherit that one. E.g. `hyperspy.misc.events_magic.py`.

vidartf committed:
Events context managers use decorator

Switched to using contextlib.contextmanager decorator instead of using
fully custom context managers.

vidartf committed:
Added `Events` black magic docstrings

vidartf committed:
Fixed Events repr

vidartf committed:
Event can be init'ed with docstring

vidartf committed:
Improve events docstrings

vidartf committed:
Add `Events.suppress_hierarchy()` + tests

vidartf committed:
Deprecate Component/Parameter.(dis)connect()

vidartf committed:
Updated AxesManager events

vidartf committed:
Made Events.children optionally callable

There are several cases where Events.children need to be populated on
the go. These changes allow it to be a callable instead of a list. Care
must naturally be taken to ensure that they are not mixed, but the
`children` attribute should only be modified in the instance that
created it.

vidartf committed:
Improved DataAxis events

vidartf committed:
AxesManager internal function rename

vidartf committed:
Deprecate AxesManager.connect/disconnect

vidartf committed:
Renamed `dummy_context_manager`

vidartf committed:
Update remaining implemented `Event`s

vidartf committed:
Various Event fixes

Various fixes from previous commits.

vidartf commented:
Implemented new proposal with all feedback taken into account.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -655,6 +720,18 @@ def _append_axis(self, *args, **kwargs):
+
+    def _on_view_changed(self):
+        self._update_attributes()
+        self.events.axes_changed.trigger()
```
Shouldn't this trigger a `view_changed` event?

vidartf replied:
Several things trigger `axes_changed` - view change is just one of them - so I don't believe it would be correct for them to correspond directly. Now there might be an additional event called `view_changed`, but i haven't yet started implementing "future" events, i.e. events that might be useful some day. I'm also not sure if those events belong in this PR?

francisco-dlp replied:
But changing `slice` actually doesn't change the axis, just the `view`, why should it trigger `axes_changed`?

vidartf replied:
What I'm trying to do here is to create an event that triggers whenever a dependent signal would need to be recomputed (other than that the data has changed). I agree that the name of the event might be wrong though. An `axes_changed` event should trigger for any change to any attribute of any `DataAxis` (including the view, and name, and unit, etc.).

A better name for this event might be `transformation_changed`? Because the transform to map the coordinates of the signal space has changed?

francisco-dlp replied:
I think that changing the view has such a dramatic impact on anything that may connect to an `AxesManager` that having a separate event for it makes sense. I don't think that triggering an `axes_changed` event makes sense here. Yes, `slice` in an attribute of `DataAxis` but the axis itself didn't change, only the way we classify it. Triggering an `axes_changed` event in this case may cause problems down the line.

I agree that currently there is no consumer for a `view_changed` event. However, I think that it makes sense to add one as we'll soon put it to good use.

vidartf replied:
There are a ton of potential useful Events in hyperspy, that's why this PR is really useful. However, I don't think its this PRs job to introduce all of them, as that will require a lot of forethought and analysis. This is the only event I'm introducing that does not replace an existing behavior in hyperspy. It might be more appropriate to remove it here, and then add it in the `interactive` PR though?

francisco-dlp replied:
I agree that this is not the time and place to come out with all the possibly useful events. However, here you have found one that will be doubtlessly useful. I don't see why we shouldn't create it at this point.

vidartf replied:
I think it is preferable to introduce new events together with a use case, and make this PR as focused as possible. It is already quite large and difficult to manage, and to be honest, I don't have the time.

francisco-dlp replied:
Ok, fair enough.
On Wed, 2016-01-13 at 07:28 -0800, Vidar Tonaas Fauske wrote:

> In hyperspy/axes.py:
> 
> > @@ -655,6 +720,18 @@ def _append_axis(self, _args, *_kwargs):
> >           axis.axes_manager = self
> >           self._axes.append(axis)
> >   
> > +    def _on_index_changed(self):
> > +        self._update_attributes()
> > +        self.events.indices_changed.trigger(axes_manager=self)
> > +
> > +    def _on_view_changed(self):
> > +        self._update_attributes()
> > +        self.events.axes_changed.trigger()
> > I think it is preferable to introduce new events together with a use
> > case, and make this PR as focused as possible. It is already quite
> > large and difficult to manage, and to be honest, I don't have the
> > time.
> > —
> > Reply to this email directly or view it on GitHub.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -469,6 +508,30 @@ class AxesManager(t.HasTraits):
     def __init__(self, axes_list):
         super(AxesManager, self).__init__()
+        self.events = Events()
+        self.events.axes_changed = Event("""
```
Why do we need an `axes_changed` event. Currently it doesn't seem to be used. Shouldn't any consumer of such an event connect to the relevant `DataAxis` event?

vidartf replied:
Sure, I will remove it, and reintroduce it in the appropriate PR where it will be used.

vidartf committed:
Remove unused events

Removed `DataAxis.axis_changed` and `AxesManager.axes_changed`, as they
are unused, and their potential use is unclear.

vidartf commented:
Updated and passed tests.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -280,7 +301,10 @@ def __deepcopy__(self, memo):
+        with self.events.index_changed.suppress():
+            self.value = self.axis[self.index]
+        self.events.index_changed.trigger(index=self.index, axis=self)
+        self.events.value_changed.trigger(value=self.index, axis=self)
```
We don't know if the value has changed, it may stay the same.

francisco-dlp commented on hyperspy/axes.py:
```diff
@@ -331,10 +355,13 @@ def index2value(self, index):
+            # If the value is above the limits we must correct the value
+            if self.continuous_value is False:
+                self.value = self.index2value(self.index)
+        self.events.index_changed.trigger(index=self.index, axis=self)
```
We don't know if `index` has actually changed.

francisco-dlp commented on hyperspy/models/model1D.py:
```diff
@@ -289,49 +292,26 @@ def update_plot(self, *args, **kwargs):
+        all_comp_param_events.children = lambda: [c.events for c in self]
+        old = self._suspend_update
+        self._suspend_update = True
+        with all_comp_param_events.suppress_hierarchy():
```
This is actually very dangerous. You are suppressing all the events of all the components in the model. Currently only the functionality that you want to pause is connected by in the future we may have other events that should not be disconnected. Also, users may connect to the event a don't expect pausing the plotting update to suppress the `value_changed` event.

francisco-dlp commented:
I think that `suppress_hierarchy` is evil. It suppress all events, which may leave the object in a dysfunctional state. Also it may suppress user created events. I think that when suppressing events we need to be as specific as possible. In particular, when possible use `Event.suppress_callback` instead of `Event.suppress`. If we need to suppress a set of events, what about something like:

``` python
es = EventSetSupressor()
es.add(event1, callback1)
es.add(event1, callback2)
es.add(event2, callback2)
es.add(event3)

with es.supress():
    do_something()

```

I think that this is more readable and customizable. I wouldn't add the feature of adding all the events in an `Events` instance. This can be done easily manually, but I think that it should be discouraged.

vidartf commented:
Hmm. I agree I might have gone a bit overboard with suppression. Especially for the `suspend_update` code. Having the ability to suspend all events on an object can be quite useful though, but I agree it should not be used (heavily) inside hyperspy itself. The usability of blanket suppression (`Events.suppress`/`suppress_hierarchy`) is to suppress all events while performing a set of changes, especially if the intermediate state is invalid (e.g. half way through changing a set of values). The correct design pattern then is to suppress all events, do the operation, (check the new state,) and then trigger manually the relevant events. It is the last step there that was forgotten in `suspend_update`.

The design you suggest seems powerful enough, so I'll implement that as well, and use it in `suspend_update`.

vidartf committed:
Added EventSupressor

vidartf committed:
Cleaning events.py

vidartf committed:
Fixed DataAxis.update_value \ set_index_from_value

vidartf commented:
Updated.

vidartf commented:
PS: I think the only "controversial" code still in this PR is the `suppress_hierarchy` function. If that is the only thing keeping this from being merged, I'll be happy to suggest it as a separate PR after the events code has been merged.

francisco-dlp commented:
`enable_plot_components` is broke. For example:

``` python

import hyperspy.api as hs
%matplotlib
import numpy as np
s = hs.signals.Spectrum(np.random.random((100, 1000)))
g1 = hs.model.components.Gaussian()
g2 = hs.model.components.Gaussian()
g1.centre.value = 400
g2.centre.value = 700
m = s.create_model()
m.extend((g1, g2))
m.plot()
m.enable_adjust_position()

```

Moving the vertical lines around raises an exception.

francisco-dlp commented:
I agree with your comment on `suppress_hierarchy`, I think that it'll be best to remove it as one can easily get the same result with `EventSupressor`. Other than that, this is brilliant!

francisco-dlp commented:
`EventSupressor` needs unittests.

vidartf committed:
Revert "Add `Events.suppress_hierarchy()` + tests"

This reverts commit 76855e8238580877d51c5976db1fac0cf313856d + following use of the implemented features.

vidartf commented:
> `enable_plot_components` is broke.

Yes, this is a consequence of the `kwargs` vs `nargs` resolution we discussed earlier. I'll try to fine tune it so it works.

vidartf commented:
Actually, strike that :)

vidartf committed:
Remove double connection

vidartf committed:
Events updates its doc with list of Events.

vidartf committed:
Improve Event auto nargs

vidartf committed:
Ensure correct parameter passing in events

vidartf committed:
Pass correct value

vidartf committed:
Use kwargs to fill in args

Since most event triggers need to specify kwargs when triggering in
order for receivers to use keywords, we rarely pass args (at least we
shouldn't in internal hyperspy code). Some functions still need to
accept them based on order limited by nargs though, so by specifying the
order as part of the Event, we can translate into args if needed.

vidartf committed:
Robust resolution of args/kwargs for nargs=all

vidartf committed:
Less strict Parameter twins

vidartf committed:
Added tests for EventsSuppressor

vidartf commented:
Pushed updates to address problems highlighted. I added some logic to argument resolution in trigger to make events as versatile and intuitive to use as possible. This makes the internals a bit more complex, but usage should be more robust. As it often is, there is a trade-off between being versatile to different usage patterns (as Events kind have to work as the glue that binds a ton of things together), and failing hard and clear on misuse.

vidartf committed:
Renamed new automatic resolution

vidartf committed:
Added tests for argument resolution

vidartf commented:
I reconsidered, and moved that automatic resolution out into a new `nargs` mode. If too much, I can pull it out.

vidartf committed:
Removed `fullauto` args resolution

Trying to streamline PR for easy acceptance.

vidartf committed:
Changed `kwarg_order` to `arguments` + black magic

Switched Event to use `arguments` instead of `kwarg_order`, and
implemented a bit of black magic to make the signature of the instance's
`trigger()` function reflect that.

vidartf committed:
Small fixes to `arguments` use

francisco-dlp commented:
Now everything seems to work. However, given how important is to get this right, I think that we need to discuss a little bit more on your solution. The actual root of the problem is that the consumer of one of the events (`_set_value`) was not designed to receive it but, with a bit of tweaking, it was possible to make it work. The ability to customize the number of args that are passed to the consumer and to pass args instead of kwargs give some extra flexibility, but at the expense of simplicity, readability and robustness. Passing args requires to know the order in which they're going to be passed and this, as previously discussed, can be a source of troubles with events. Wouldn't it be better to simply remove the ability to pass args and to customize their number? Where a function can not be written on purpose for an event, it can always be wrapped.

vidartf commented:
Regarding the previous case that didn't work, that was simply an issue of me not completely understanding how `twins` in `Parameter` worked, especially when a `DataAxis` was duck-typed as a twin. All the other changes was simply because I identified some weaknesses in the old implementation.

The input side is now very robust, and I definitely think we should keep that functionality (I'm referring to initialization arg `arguments` giving `trigger()` a signature).

On the output side, I've gone for the plug-and-play option. There are very few cases that currently don't work out of the box. If any of those cases should ever occur in the code, those can of course be wrapped, but from my point of view, wrapping should be the exception, not the default. And simply adding `**kwargs` to every function that should be connected is not ideal, as it will hide any `unexpected keyword argument` errors for those functions in ordinary use.

Also, as you can see from the cases were events are already used, in almost all cases I've found using `nargs` to be the simplest and most robust way of connecting things.

I do agree the internals of `nargs` trigger resolution now are a bit complex, but as far as I see it, that is just complexity that has been removed from the rest of the code (and 10 lines of complex code in one place is better than 30 opaque lines of code spread throughout the codebase). Can you give an example where events currently are not simple, readable and robust, from the user's point of view? If there are still any major sources of confusion, or possible errors, we might have to scrap the current implementation. However, if it works as it is now it would be pretty advantageous.

francisco-dlp commented:
I wasn't meaning the complexity of the internal code, but the added complexity of having several ways of accomplishing the same task. I don't like that things can get broken by not getting the arguments order right. What about explicitly mapping event kwargs to function kwargs e.g.:

``` python
index_changed.connect(f, ["axis"]) # pass just the axis kwarg
index_changed.connect(f) # pass all events
index_changed.connect(f, ["axis" : "ax", "index": "idx"]) # Map the event kwargs to match those in the function
index_changed.connect(f, []) # don't pass any event

```

I think that this is more explicit and robust.

vidartf commented:
The code you suggest is definitely more verbose, but requires that the code that connects the event and function has complete knowledge of the signature of both the event and function (unless its a 0 `nargs` case). An example of the 0 nargs case is the events that get used in interactive.

I think it might be easiest if you simply merge this, and modify it as you want it, then I can review the changes on the main PR, as I'm out of time I can spend on this.

vidartf commented:
PS: A scenario for comparing the suggestions (maybe relevant for widgets):

Say you have a class A, that connects its function `do_something()` to an external event. The external event has some arguments, but `do_something` doesn't need them. It therefore connects with `nargs = 0`. Then later on, class B inherits class A, and overrides `do_something(arg1, arg2=None)`. If this function then wants some of the arguments, there are a some changes that might need doing:
1. With the current implementation, and `nargs = 'auto'`, everything is OK, as long as the event supplies both `arg1` and `arg2` (the 'fullauto' mode that I implemented would work either way).
2. With the current implementation, and `nargs = 'all'`, or your suggestion where all is passed, `A.do_something()` would need `**kwargs`, and so would B's variant (if it doesn't name all supplied by event).
3. With the current implementation and `nargs = X`, or the implementation you suggest with specified keywords, class B would have to override class A's connection code (so they shouldn't be in init).

francisco-dlp commented:
> The code you suggest is definitely more verbose, but requires that the code that connects the event and function has complete knowledge of the signature of both the event and function (unless its a 0 `nargs` case). An example of the 0 nargs case is the events that get used in interactive.

Yes, but that's hardly a bad thing. It fixes the issue of connecting to functions that are not explicitly designed to receive the event in a more general way and customizable way. This should completely remove the need for wrapping.

> I think it might be easiest if you simply merge this, and modify it as you want it, then I can review the changes on the main PR, as I'm out of time I can spend on this.

That's fair enough. I'll merge it.

Regarding the "auto" mode of the current implementation, I agree that it can be useful in some corner cases. But there is no reason why we can't keep it:

``` python
index_changed.connect(f, "auto")

```
